File: app/src/main/java/akiniyalocts/imgurapiexample/imgurmodel/ImgurAPI.java
package akiniyalocts.imgurapiexample.imgurmodel;

import retrofit.http.Body;
import retrofit.http.Header;
import retrofit.http.POST;
import retrofit.http.Query;
import retrofit.mime.TypedFile;

/**
 * Created by AKiniyalocts on 2/23/15.
 */
public interface ImgurAPI {

  /****************************************
   * Upload
   * Image upload API
   */

  /**
   *
   * @param auth #Type of authorization for upload
   * @param title #Title of image
   * @param description #Description of image
   * @param albumId #ID for album (if the user is adding this image to an album)
   * @return
   */
  @POST("/3/image")ImageResponse postImage (
      @Header("Authorization")String auth,
      @Query("title")String title,
      @Query("description")String description,
      @Query("album")String albumId,
      @Query("account_url")String username,
      @Body TypedFile file
  );
}

File: app/src/main/java/com/akiniyalocts/imgurapiexample/api/ImgurApi.kt
package com.akiniyalocts.imgurapiexample.api

import okhttp3.MultipartBody
import okhttp3.ResponseBody
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.Part

interface ImgurApi {
    @Multipart
    @POST("/3/upload")
    suspend fun uploadFile(
        @Part("image") image: MultipartBody.Part?,
        @Part("video") video: MultipartBody.Part?,
        @Part("album") album: MultipartBody.Part? = null,
        @Part("type") type: MultipartBody.Part,
        @Part("name") name: MultipartBody.Part?,
        @Part("title") title: MultipartBody.Part?,
        @Part("description") description: MultipartBody.Part?,
        @Part("disable_audio") disableAudio: Boolean? = null
    ): Result<ResponseBody>
}